File: hotel-cn334/backend/schemas/booking.py
import re
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, field_validator
from pydantic.alias_generators import to_camel


class GuestCount(BaseModel):
    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)

    adults: int
    children: int


class DateRange(BaseModel):
    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)

    check_in: str
    check_out: str


class ContactInfo(BaseModel):
    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)

    title: str
    full_name: str
    phone_number: str
    email: str

    @field_validator("email")
    @classmethod
    def validate_email(cls, v: str) -> str:
        if not re.match(r"\S+@\S+\.\S+", v):
            raise ValueError("Invalid email format")
        return v


class BookingCreate(BaseModel):
    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)

    room_type: str
    price: int
    dates: DateRange
    guests: GuestCount
    contact: ContactInfo
    payment_method: Literal["credit_card", "thai_qr", "alipay", "wechat"]


class BookingResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)

    booking_id: str
    status: str
    message: str
    data: Optional[BookingCreate] = None

File: hotel-cn334/frontend/types/booking.ts
// Booking Data Interfaces
export interface GuestCount {
  adults: number;
  children: number;
}

export interface DateRange {
  checkIn: string;
  checkOut: string;
}

export interface ContactInfo {
  title: string;
  fullName: string;
  phoneNumber: string;
  email: string;
}

export interface PaymentMethod {
  type: 'credit_card' | 'thai_qr' | 'alipay' | 'wechat';
  label: string;
}

export interface BookingFormData {
  roomType: string;
  price: number;
  dates: DateRange;
  guests: GuestCount;
  contact: ContactInfo;
  paymentMethod: string;
}

export interface BookingResponse {
  bookingId: string;
  status: 'success' | 'pending' | 'failed';
  message: string;
  data?: BookingFormData;
}

export interface QRCodeData {
  bookingId: string;
  roomType: string;
  checkIn: string;
  checkOut: string;
  guestName: string;
}

export type BookingStep = 'form' | 'qr' | 'success' | 'error';
